jax/experimental/host_callback.py

```diff
@@ -0,0 +1,797 @@
+                                     **_add_transform_name(params, "jvp"))
+  return tuple(out_primals), tuple(out_tangents_extra[:-1])
+
+
+ad.primitive_jvps[id_tap_p] = _id_tap_jvp_rule
+
+
+def _id_tap_transpose_rule(cts, *args, func=None, nr_untapped=0, **params):
+  assert len(cts) == len(args)
+  cts_zeros = tuple(map(_instantiate_zeros, args, cts))
+  ct_args = id_tap_p.bind(*cts_zeros, func=func, nr_untapped=nr_untapped,
+                          **_add_transform_name(params, "transpose"))
+  return ct_args
+
+
+ad.primitive_transposes[id_tap_p] = _id_tap_transpose_rule
+
+
+def _id_tap_batching_rule(batched_args, batch_dims, **params):
+  new_params = _add_transform_name(params, "batch")
+  new_params["batch_dims"] = batch_dims
+  res = id_tap_p.bind(*batched_args, **new_params)
+  return res, batch_dims
+
+
+batching.primitive_batchers[id_tap_p] = _id_tap_batching_rule
+
+# def _id_tap_shape_rule(*operands, **params):
+#  return tuple([op.shape for op in operands])
+
+# TODO: these disappeared
+# masking.shape_rules[id_tap_p] = _id_tap_shape_rule  # type: ignore[module-attr]
+
+def _id_tap_masking_rule(operands, operands_logical_shapes, **params):
+  new_params = _add_transform_name(params, "mask")
+  new_params["logical_shapes"] = operands_logical_shapes
+  return id_tap_p.bind(*operands, **new_params)
+
+
+masking.masking_rules[id_tap_p] = _id_tap_masking_rule
+
+#### XLA compilation ####
+####
+# Special consumer to mark the end of outfeed stream for a device
+_end_consumer = 0
+_unknown_consumer = 1  # for testing error cases
+
+
+def _id_print_translation_rule_outfeed(comp: XlaComputationBuilder,
+                                       *args_op: XlaOp, func=None,
+                                       nr_untapped=0, arg_treedef=None,
+                                       **params):
+  params = dict(params)
+  if func is _end_consumer:
+    params["consumer_id"] = _end_consumer
+  elif func is _unknown_consumer:
+    params["consumer_id"] = _unknown_consumer  # Will trigger an error, for testing
+  else:
+    params["consumer_id"] = _register_consumer(
+      _ConsumerCallable(func, tuple(params.items()), arg_treedef))
+
+
+  # We expect the current token at the end
+  current_token = args_op[-1]
+  current_token_shape = comp.GetShape(current_token)
+  if current_token_shape.is_array():
+    # TODO: we get here because wehn we partially eval some primitives
+    # we impl themn with JIT, but we did not rewrite them
+    has_token = False
+    current_token = xops.CreateToken(comp)
+  else:
+    has_token = True
+
+  nr_args_to_emit = len(args_op) - nr_untapped - (1 if has_token else 0)
+  next_token = _emit_outfeed(comp, current_token,
+                             args_op[0:nr_args_to_emit], params["consumer_id"])
+  results = (args_op[:-1] + (next_token,)) if has_token else args_op
+  return xops.Tuple(comp, results)
+
+
+xla.translations[id_tap_p] = _id_print_translation_rule_outfeed
```

Comment:
Should we rename `_id_print_translation_rule_outfeed` -> `_id_tap_translation_rule_outfeed`?

Reply:
Done